### ISSUE TITLE
Add support for per thing channel permissions

### ISSUE THREAD

**Issue by Thalhammer** [absmach/magistrala#1013]
**FEATURE REQUEST**

1. Is there an open issue addressing this request? If it does, please add a "+1" reaction to the
   existing issue, otherwise proceed to step 2.

Did not find one.

2. Describe the feature you are requesting, as well as the possible use case(s) for it.
Maybe I don't get the correct usage of this plattform but i would like to limit device access to each channel on a per thing (or at least thing type). I.e. some things can write (publish) to the channel, some can only read (subscribe), some can do both.

3. Indicate the importance of this feature to you (must-have, should-have, nice-to-have).
There are a lot of potential use cases, e.g.
* Firmware update (A device should not be able to publish to a channel used to announce firmware updates)
* Device separation (Device A should be able to subscribe to measurements of device B).

Currently the only way to achieve this kind of setup is by using an individual channel for each device, which is probably not the correct thing to do.
Please correct me if I am wrong, but to me this feels like a really fundamental feature.

**Comment by drasko** [comment]
> Maybe I don't get the correct usage of this plattform but i would like to limit device access to each channel on a per thing (or at least thing type). I.e. some things can write (publish) to the channel, some can only read (subscribe), some can do both.

This makes sense and is on the roadmap with policies for the messaging: https://github.com/mainflux/mainflux/pull/960

> - Firmware update (A device should not be able to publish to a channel used to announce firmware updates)

Yes, this use-case makes sense: device would connect to channel in read-only mode, while application that triggers update will connect to channel in write mode.

> - Device separation (Device A should be able to subscribe to measurements of device B).

This is possible now, just that once connected to the same channel they will be capable to both write on it.

> Currently the only way to achieve this kind of setup is by using an individual channel for each device, which is probably not the correct thing to do.

This is wrong - you can connect unlimited number of devices on the same channel. Just that they will be all capable to write (publish), not only listen (subscribe).

> Please correct me if I am wrong, but to me this feels like a really fundamental feature.

I do not this that this is so fundamental from the following reasons:
- Only user who provisions the whole solution (and who creates devices and channels) can "connect" certain devices to channels (as they belong only to him)
- This person is system administrator and probably the creator of the firmware for the devices
- He will make a firmware that on certain channels device A acts as a publisher and all other devices as subscribers only (in the end, channels are topics, and what which device will do on which topic must be configured in the firmware)

Access control policies is something that is good to have, but it is really not fundamental for creating secure system with Mainflux, as all use-cases that we saw so far fall into this category - person who provisions devices and channels actually controls the firmware itself.

Let's think about what would happen if this was not the case - you would provision the system and then give credentials to some device to which you are not certain what it will do - it can publish, subscribe, do arbitrary stuff. How realistic is that? Not so, I would say.

Maybe I am seeing this wrong, and I would like to understand your use-case better, where you have some potentially malicious devices, that you are not sure what their firmware is doing - but you did give them credentials or X.509 certificates and access to your system anyway.

**Comment by nmarcetic** [comment]
@Thalhammer As @drasko said PR #960 will solve everything you mentioned. It will be finished and merged  in the next few weeks.

**Comment by Thalhammer** [comment]
> This is possible now, just that once connected to the same channel they will be capable to both write on it.

Sorry it looks like I forgot a "not" in my original issue. "Device A should not be able to subscribe to measurements of device B".

> Access control policies is something that is good to have, but it is really not fundamental for creating secure system with Mainflux, as all use-cases that we saw so far fall into this category - person who provisions devices and channels actually controls the firmware itself.

I highly disagree. You don't control the firmware anymore at the point it leaves your hands. There is simply no current technology stoping someone with enough interest from gaining access to passwords and tokens inside the device. It started correctly (using per device user & passwords) but being unable to restrict a device from writing to stuff it should only be able to read kind of destroys the whole point of it. As you can simply extract the username and password, connect to the server and publish a new and shiny firmware including your super cool virus, malware, whatever.

> This makes sense and is on the roadmap with policies for the messaging: #960

I did not see this PR, and once it is merged my issue should be gone, however I just wanted to note my point that you can never ever trust a device (unless it never left your hand) even if you (think you) control the firmware. Because in reality you don't.
